### PR TITLE
Added asynchronous buttons that will be disabled during the action

### LINKED ISSUE #216: Preventing double clicks during asynchronous actions
As a user I'd like to have double clicks prevented to avoid performing duplicate actions

### PR DESCRIPTION
**Pull request by terolamminpaa**
resolves #216

### CODE REVIEW

File: src/components/contracts/ContractList.tsx [hunk 1]
```diff
@@ -269,9 +270,9 @@ class ContractList extends React.Component<Props, State> {
             contractDatas={this.state.freshContractsPast}
           />
         }
-        <Button onClick={() => this.onProposeNewContractClick("FRESH")} style={{ borderRadius: 0 }} color="red">
+        <AsyncButton onClick={ async () => await this.onProposeNewContractClick("FRESH")} style={{ borderRadius: 0 } } color="red">
```
Comment (lauriLukkarinen) [inline src/components/contracts/ContractList.tsx:273]:
```suggestion
        <AsyncButton
          onClick={ async () => await this.onProposeNewContractClick("FRESH") }
          style={{ borderRadius: 0 } }
          color="red"
        >
```

File: src/components/deliveries/IncomingDeliveries.tsx [hunk 2]
```diff
@@ -352,7 +353,7 @@ class IncomingDeliveries extends React.Component<Props, State> {
           <Modal.Actions>
             <Button.Group>
               <Button onClick={() => this.setState({ confirmBeginDelivery: false })} color="red" inverted>Sulje</Button>
-              <Button onClick={() => this.handleBeginDelivery(this.state.deliveryProduct)} positive icon='check' labelPosition='right' content='Aloita toimitus' />
+              <AsyncButton onClick={async () => await this.handleBeginDelivery(this.state.deliveryProduct)} positive icon='check' labelPosition='right' content='Aloita toimitus' />
```
Comment (lauriLukkarinen) [inline src/components/deliveries/IncomingDeliveries.tsx:356]:
```suggestion
              <AsyncButton
                onClick={async () => await this.handleBeginDelivery(this.state.deliveryProduct) }
                positive
                icon='check'
                labelPosition='right'
                content='Aloita toimitus'
              />
```

File: src/components/deliveries/IncomingDeliveries.tsx [hunk 3]
```diff
@@ -364,7 +365,7 @@ class IncomingDeliveries extends React.Component<Props, State> {
           <Modal.Actions>
             <Button.Group>
               <Button onClick={() => this.setState({ confirmRemove: false })} color="red" inverted>Sulje</Button>
-              <Button onClick={() => this.handleRemoveDelivery(this.state.deliveryProduct)} color="red" icon='trash' labelPosition='right' content='Hylkää toimitus' />
+              <AsyncButton onClick={async () => await this.handleRemoveDelivery(this.state.deliveryProduct)} color="red" icon='trash' labelPosition='right' content='Hylkää toimitus' />
```
Comment (lauriLukkarinen) [inline src/components/deliveries/IncomingDeliveries.tsx:368]:
```suggestion
              <AsyncButton
                onClick={async () => await this.handleRemoveDelivery(this.state.deliveryProduct) }
                color="red"
                icon='trash'
                labelPosition='right'
                content='Hylkää toimitus'
              />
```

File: src/components/deliveries/ManageDeliveryModal.tsx [hunk 4]
```diff
@@ -578,7 +579,7 @@ class ManageDeliveryModal extends React.Component<Props, State> {
                         <p style={{ padding: 20 }}> {deliveryNote.text}</p>
                       </div>
                       <div style={{ display: "flex", flex: 0.3, minHeight: "100px", alignItems: "center" }}>
-                        <Button onClick={() => this.removeNote(deliveryNote, i)} color="black">Poista huomio</Button>
+                        <AsyncButton onClick={async () => await this.removeNote(deliveryNote, i)} color="black">Poista huomio</AsyncButton>
```
Comment (lauriLukkarinen) [inline src/components/deliveries/ManageDeliveryModal.tsx:582]:
```suggestion
                        <AsyncButton onClick={ async () => await this.removeNote(deliveryNote, i) } color="black">Poista huomio</AsyncButton>
```

File: src/components/generic/asynchronous-button.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * Component render
+   */
+  public render = () => {
+    const { children, disabled } = this.props;
+    const { loading } = this.state
+
+    return (
+      <Button { ...this.props } disabled={ loading ?? disabled } loading={ loading } onClick={ this.handleClick }>
```
Comment (lauriLukkarinen) [inline src/components/generic/asynchronous-button.tsx:41]:
```suggestion
      <Button
        { ...this.props }
        disabled={ loading ?? disabled }
        loading={ loading }
        onClick={ this.handleClick }
      >
```

File: src/components/generic/asynchronous-button.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  /**
+   * Method for handling button press
+   */
+  private handleClick = () => {
+    this.handleAsync();
+  }
+
+  private handleAsync = async () => {
```
Comment (lauriLukkarinen) [inline src/components/generic/asynchronous-button.tsx:54]: Check if these can be combined to just one async handleClick method, button component should accept that

File: src/components/itemgroups-management/CreateAndEditItemGroupPriceModal.tsx [hunk 7]
```diff
@@ -228,7 +229,7 @@ class CreateAndUpdateItemGroupPriceModal extends React.Component<Props, State> {
           <Button.Group floated="right" className="modal-button-group" >
             <Button onClick={this.closeModal} color="black">Sulje</Button>
             <Button.Or text="" />
-            <Button onClick={this.state.edit ? this.handleUpdatePrice : this.handleCreatePrice} color="red">{this.state.edit ? "Tallenna muutokset" : "Lisää uusi hinta"}</Button>
+            <AsyncButton onClick={this.state.edit ? this.handleUpdatePrice : this.handleCreatePrice} color="red">{this.state.edit ? "Tallenna muutokset" : "Lisää uusi hinta"}</AsyncButton>
```
Comment (lauriLukkarinen) [inline src/components/itemgroups-management/CreateAndEditItemGroupPriceModal.tsx:232]:
```suggestion
            <AsyncButton
              onClick={ this.state.edit ? this.handleUpdatePrice : this.handleCreatePrice }
              color="red"
            >
              { this.state.edit ? "Tallenna muutokset" : "Lisää uusi hinta" }
            </AsyncButton>
```

File: src/components/itemgroups-management/ItemGroupPricesList.tsx [hunk 8]
```diff
@@ -186,7 +187,7 @@ class ItemGroupPricesList extends React.Component<Props, State> {
                       <Button.Group className="contract-button-group" >
                         <Button onClick={() => this.setState({ edit: true, itemGroupPriceId: itemGroupPrice.id, createModal: true })} color="red">Muokkaa</Button>
                         <Button.Or text="" />
-                        <Button onClick={() => this.handleDelete(itemGroupPrice.id || "")} color="black">Poista</Button>
+                        <AsyncButton onClick={async () => await this.handleDelete(itemGroupPrice.id || "")} color="black">Poista</AsyncButton>
```
Comment (lauriLukkarinen) [inline src/components/itemgroups-management/ItemGroupPricesList.tsx:190]:
```suggestion
                        <AsyncButton onClick={ async () => await this.handleDelete(itemGroupPrice.id || "") } color="black">Poista</AsyncButton>
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 9]
```diff
@@ -681,11 +682,17 @@ class OpeningHoursScreen extends React.Component<Props, State> {
             <Form>
                 <Form.Field style={{ display: "flex", marginBottom: "2rem", marginRight: "1.5rem" }}>
                   <Header style={{ marginRight: "1rem" }}>{ this.formatDate(openingHourException.exceptionDate) }</Header>
-                  <Button style={{ fontSize: "0.8rem" }} color="red" onClick={ this.deleteExceptionHoursBlock(openingHourException.id) }>{ strings.deleteBlock }</Button>
+                  <AsyncButton style={{ fontSize: "0.8rem" }} color="red" onClick={ async () => { await this.deleteExceptionHoursBlock(openingHourException.id) } }>{ strings.deleteBlock }</AsyncButton>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:685]:
```suggestion
                  <AsyncButton
                    style={{ fontSize: "0.8rem" }}
                    color="red"
                    onClick={ async () => { await this.deleteExceptionHoursBlock(openingHourException.id) } }
                  >
                    { strings.deleteBlock }
                  </AsyncButton>
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 10]
```diff
@@ -681,11 +682,17 @@ class OpeningHoursScreen extends React.Component<Props, State> {
...
                 </Form.Field>
                 <Form.Field>
-                  <span style={{ cursor: "pointer" }} onClick={ this.deleteExceptionHours(openingHourException, openingHourException.id) }>{ strings.deleteHours }<Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="trash" /></span>
-                  <span style={{ cursor: "pointer" }} onClick={ this.addExceptionHours(openingHourException, openingHourException.id) }>{ strings.addNewHours }<Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="add circle" /></span>
+                  <AsyncButton style={{ background: "transparent" }} onClick={ async () => { await this.deleteExceptionHours(openingHourException, openingHourException.id) } }>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:688]:
```suggestion
                  <AsyncButton
                    style={{ background: "transparent" }}
                    onClick={ async () => await this.deleteExceptionHours(openingHourException, openingHourException.id) }
                  >
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 11]
```diff
@@ -681,11 +682,17 @@ class OpeningHoursScreen extends React.Component<Props, State> {
...
+                    { strings.deleteHours }
+                    <Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="trash" />
+                  </AsyncButton>
+                  <AsyncButton style={{ background: "transparent" }} onClick={ async () => { await this.addExceptionHours(openingHourException, openingHourException.id) } }>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:692]:
```suggestion
                  <AsyncButton
                    style={{ background: "transparent" }}
                    onClick={ async () => await this.addExceptionHours(openingHourException, openingHourException.id) }
                  >
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 12]
```diff
@@ -916,9 +924,15 @@ class OpeningHoursScreen extends React.Component<Props, State> {
             {
               openingHourWeekday.hours.map((hours, index) => this.renderHoursRow(openingHourPeriod, index, hours, false, openingHourWeekday))
             }
-            <Form.Field>
-              <span style={{ cursor: "pointer" }} onClick={ this.deleteOpeningHours(openingHourWeekday, openingHourPeriod) }>{ strings.deleteHours }<Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="trash" /></span>
-              <span style={{ cursor: "pointer" }} onClick={ this.addOpeningHours(openingHourWeekday, openingHourPeriod) }>{ strings.addNewHours }<Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="add circle" /></span>
+            <Form.Field style={{ display: "flex", flexDirection: "row" }}>
+              <AsyncButton style={{ background: "transparent", flex: 1, padding: 0 }} onClick={ async () => { this.deleteOpeningHours(openingHourWeekday, openingHourPeriod) } }>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:928]:
```suggestion
              <AsyncButton
                style={{ background: "transparent", flex: 1, padding: 0 }}
                onClick={ async () => this.deleteOpeningHours(openingHourWeekday, openingHourPeriod) }
              >
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 13]
```diff
@@ -916,9 +924,15 @@ class OpeningHoursScreen extends React.Component<Props, State> {
...
+                { strings.deleteHours }
+                <Icon style={{ marginRight: "1rem", marginLeft: "1rem" }} name="trash" />
+              </AsyncButton>
+              <AsyncButton style={{ background: "transparent", flex: 1, padding: 0 }} onClick={ async () => { this.addOpeningHours(openingHourWeekday, openingHourPeriod) } }>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:932]:
```suggestion
              <AsyncButton
                style={{ background: "transparent", flex: 1, padding: 0 }}
                onClick={ async () => this.addOpeningHours(openingHourWeekday, openingHourPeriod) }
              >
```

File: src/components/qualities/CreateAndEditQuality.tsx [hunk 14]
```diff
@@ -223,9 +224,9 @@ class CreateAndEditQuality extends React.Component<Props, State> {
               inverted
               color="red">Takaisin</Button>
             <Button.Or text="" />
-            <Button color="red" disabled={!this.isValid()} onClick={() => this.state.selectedDeliveryQualityId === "new" ? this.handleCreateClick() : this.handleUpdateClick()} type='submit'>
+            <AsyncButton color="red" disabled={!this.isValid()} onClick={ this.state.selectedDeliveryQualityId === "new" ? this.handleCreateClick : this.handleUpdateClick } type='submit'>
```
Comment (lauriLukkarinen) [inline src/components/qualities/CreateAndEditQuality.tsx:227]:
```suggestion
            <AsyncButton
              color="red"
              disabled={ !this.isValid() }
              onClick={ this.state.selectedDeliveryQualityId === "new" ? this.handleCreateClick : this.handleUpdateClick }
              type='submit'
            >
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/generic/asynchronous-button.tsx (hunk 6 above)
Comment (terolamminpaa) [reply src/components/generic/asynchronous-button.tsx]: You are right. This was probably only necessary on the app side with react-native buttons. Fixed to just one function.

Comment (terolamminpaa) [reply src/components/generic/asynchronous-button.tsx]: I was wrong. It isn't required in there either

File: src/components/deliveries/IncomingDeliveries.tsx [hunk 15]
```diff
@@ -352,7 +353,13 @@ class IncomingDeliveries extends React.Component<Props, State> {
           <Modal.Actions>
             <Button.Group>
               <Button onClick={() => this.setState({ confirmBeginDelivery: false })} color="red" inverted>Sulje</Button>
-              <Button onClick={() => this.handleBeginDelivery(this.state.deliveryProduct)} positive icon='check' labelPosition='right' content='Aloita toimitus' />
+              <AsyncButton
+                onClick={async () => await this.handleBeginDelivery(this.state.deliveryProduct) }
```
Comment (lauriLukkarinen) [inline src/components/deliveries/IncomingDeliveries.tsx:357]:
```suggestion
                onClick={ async () => await this.handleBeginDelivery(this.state.deliveryProduct) }
```

File: src/components/deliveries/IncomingDeliveries.tsx [hunk 16]
```diff
@@ -364,7 +371,13 @@ class IncomingDeliveries extends React.Component<Props, State> {
           <Modal.Actions>
             <Button.Group>
               <Button onClick={() => this.setState({ confirmRemove: false })} color="red" inverted>Sulje</Button>
-              <Button onClick={() => this.handleRemoveDelivery(this.state.deliveryProduct)} color="red" icon='trash' labelPosition='right' content='Hylkää toimitus' />
+              <AsyncButton
+                onClick={async () => await this.handleRemoveDelivery(this.state.deliveryProduct) }
```
Comment (lauriLukkarinen) [inline src/components/deliveries/IncomingDeliveries.tsx:375]:
```suggestion
                onClick={ async () => await this.handleRemoveDelivery(this.state.deliveryProduct) }
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 17]
```diff
@@ -565,7 +566,7 @@ class OpeningHoursScreen extends React.Component<Props, State> {
             endDate={ openingHourPeriod.endDate }
           />
         </Form.Field>
-        <Form.Field onClick={ this.deleteOpeningHoursBlock(openingHourPeriod, openingHourPeriod.id) } style={{ cursor: "pointer" }}>
+        <Form.Field onClick={ async () => { this.deleteOpeningHoursBlock(openingHourPeriod, openingHourPeriod.id) } } style={{ cursor: "pointer" }}>
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:569]:
```suggestion
        <Form.Field onClick={ async () => this.deleteOpeningHoursBlock(openingHourPeriod, openingHourPeriod.id) } style={{ cursor: "pointer" }}>
```

File: src/components/opening-hours-management/OpeningHoursScreen.tsx [hunk 18]
```diff
@@ -681,11 +682,29 @@ class OpeningHoursScreen extends React.Component<Props, State> {
             <Form>
                 <Form.Field style={{ display: "flex", marginBottom: "2rem", marginRight: "1.5rem" }}>
                   <Header style={{ marginRight: "1rem" }}>{ this.formatDate(openingHourException.exceptionDate) }</Header>
-                  <Button style={{ fontSize: "0.8rem" }} color="red" onClick={ this.deleteExceptionHoursBlock(openingHourException.id) }>{ strings.deleteBlock }</Button>
+                  <AsyncButton
+                    style={{ fontSize: "0.8rem" }}
+                    color="red"
+                    onClick={ async () => { await this.deleteExceptionHoursBlock(openingHourException.id) } }
```
Comment (lauriLukkarinen) [inline src/components/opening-hours-management/OpeningHoursScreen.tsx:688]:
```suggestion
                    onClick={ async () => await this.deleteExceptionHoursBlock(openingHourException.id) }
```

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**